### PR TITLE
✨ NEW: Keep link references

### PR DESCRIPTION
**Pull request by chrisjsewell**
Utilise the new `store_labels` option I introduced in mardown-it-py, to keep link/image reference labels, instead of their resolved href. Then output the (sorted) references at the end of the document.

Notes:
- I've left a "hook" to turn this feature on/off: `markdown_it.options["mdformat"] = {"keep_references": True}`, this could be turned into a CLI option, if desired.
- References are always normalized using `str.upper()` in markdown-it, but, as a style choice, I convert them to `str.lower()`
- I've added a `finalize` key-word argument to `MDRenderer.render`. This is for use by plugins (like mdformat-tables) that need to run "nested rendering", during which they do not want these references output.

### CODE REVIEW

**Comment by chrisjsewell** [comment]
FYI, whats the reasoning for always wrapping link URLs in `<>`? https://github.com/executablebooks/mdformat/blob/200aaea3d408582c65fdeffd60f46822e8436495/mdformat/renderer/_token_renderers.py#L37

**Comment by hukkinj1** [comment]
> Also, whats the reasoning for always wrapping link URIs in <>? ... If I recall correctly, this signifies an external link

I don't think CommonMark spec makes a distinction between external or internal links. With the `<>` brackets the spec allows a wider set of characters and less character escaping than without. So the formatting logic is a lot simpler with the brackets, and all links look the same, those are the reasons for the wrapping.

**Comment by hukkinj1** [comment]
> I've left a "hook" to turn this feature on/off: markdown_it.options["mdformat"] = {"keep_references": True}, this could be turned into a CLI option, if desired.

As a bit of a minimalist, I would prefer to remove this hook. Let's just change the default style to "keep ref links as ref links and inline links as inline links".

> References are always normalized using str.upper() in markdown-it, but, as a style choice, I convert them to str.lower()

Very good, i like this.

**Comment by hukkinj1** [comment]
I would prefer mdformat to delete unused refs (similar to how it deletes duplicate refs). Might want to make a test case that checks that none of these "extra newlines before references" are left at the end of file if there's zero refs.

File: mdformat/renderer/__init__.py [hunk 1]
```diff
@@ -99,8 +105,27 @@ def render(
...
+            ):
+                rendered_content += "\n\n"
+                rendered_content += self.write_references(env)
+
             rendered_content += "\n"
+
         return rendered_content
+
+    @staticmethod
+    def write_references(env: dict) -> str:
```
Comment (hukkinj1) [inline mdformat/renderer/__init__.py:123]: Does this need to be exported? I'd prefer it to be private method if not.

File: tests/test_renderer_style.py [hunk 2]
```diff
@@ -1,45 +1,18 @@
-from markdown_it import MarkdownIt
+from pathlib import Path
+
+from markdown_it.utils import read_fixture_file
 import pytest
 
-from mdformat.renderer import MDRenderer
+from mdformat import text as render
```
Comment (hukkinj1) [inline tests/test_renderer_style.py:6]: What's the reason for this change? I would perhaps prefer to leave this as a test for the `renderer` package. There is another one for the public API functions.

File: tests/test_renderer_style.py [hunk 3]
```diff
@@ -1,45 +1,18 @@
...
+    "line,title,text,expected", STYLE_CASES, ids=[f[1] for f in STYLE_CASES]
+)
+def test_renderer_style(line, title, text, expected):
     """Test Markdown renderer renders expected style."""
-    md_original = entry["input_md"]
-    md_new = MarkdownIt(renderer_cls=MDRenderer).render(md_original)
-    expected_md = entry["output_md"]
-    assert md_new == expected_md
+    md_new = render(text)
+    print(md_new)
```
Comment (hukkinj1) [inline tests/test_renderer_style.py:17]: Should we remove the print?

**Comment by chrisjsewell** [comment]
> I don't think CommonMark spec makes a distinction between external or internal links. With the <> brackets the spec allows a wider set of characters and less character escaping than without. So the formatting logic is a lot simpler with the brackets, and all links look the same, those are the reasons for the wrapping.

As with #33 I suggest these type of considerations are documented, since I'm sure I won't be the only one question this.

I guess then that this should be the same for image uri?

File: tests/test_renderer_style.py (hunk 3 above)
Comment (chrisjsewell) [reply tests/test_renderer_style.py:17]: I'm going to leave it in, but only when the assertion fails, since it's a very handy practice for adding/modifying tests, to add a new test input, run pytest, then copy/paste the printed output to the expected section.

File: tests/test_renderer_style.py (hunk 2 above)
Comment (chrisjsewell) [reply tests/test_renderer_style.py:6]: I've changed it pack to using the renderer, but note you have to add:

```python
md_new.options["store_labels"] = True
```

for the references to be recovered (see https://github.com/executablebooks/markdown-it-py/commit/d9b6de9c8ff3bb0dc2b78d6eda8e09b3b88fa7b4)

File: mdformat/renderer/__init__.py (hunk 1 above)
Comment (chrisjsewell) [reply mdformat/renderer/__init__.py:123]: Why not allow it to be overridden in a subclass?
In fact, technically I'd say it's a bit of a shortcoming of this renderer; that you can't override any of the builtin token render methods in a subclass, as opposed to https://github.com/executablebooks/markdown-it-py/blob/master/markdown_it/renderer.py where the render method for every token can be overridden.
There's not an immediate use case for that though, and you can now achieve it via plugins, so I'm not suggesting changing it, just noting 😉

**Comment by chrisjsewell** [comment]
> As a bit of a minimalist, I would prefer to remove this hook. Let's just change the default style to "keep ref links as ref links and inline links as inline links".

done

> I would prefer mdformat to delete unused refs (similar to how it deletes duplicate refs). 

done

> Might want to make a test case that checks that none of these "extra newlines before references" are left at the end of file if there's zero refs.

new lines and references are only added if there are any referenced labels, so this is not an issue

File: tests/test_renderer_style.py (hunk 2 above)
Comment (hukkinj1) [reply tests/test_renderer_style.py]: Ah alright. What happens if `store_labels = False` now? The ref links just get formatted as inline (which would seem fine). If it's something more unexpected, we might want to just fail in `MDRenderer.__init__` with a ValueError if store_labels isnt enabled.

Comment (chrisjsewell) [reply tests/test_renderer_style.py]: yep they just get formatted as inline 👍

**Comment by hukkinj1** [comment]
> As with #33 I suggest these type of considerations are documented, since I'm sure I won't be the only one question this.

:+1: made an issue for this

> I guess then that this should be the same for image URIs?

I would imagine so as well. I'm pretty sure there's some test that breaks if the angle brackets are removed from normal links. We should probably make similar tests as the failing ones, but for images, and see what happens. Making an issue.

File: mdformat/renderer/__init__.py (hunk 1 above)
Comment (hukkinj1) [reply mdformat/renderer/__init__.py:121]: So by default I always want to make everything private to avoid a maintenance nightmare. Any changes, refactoring etc. that change public API is a breaking change and forces me to make a major version bump, and is a maintenance hassle for a potential user.

> In fact, technically I'd say it's a bit of a shortcoming of this renderer;

It is a shortcoming for someone willing to subclass the class. It is also a major advantage for me trying to maintain the project. So far someone subclassing the renderer class is hypothetical, but me maintaining it is reality.

I didn't really intend to make even `MDRenderer.render` public API, it just happened to be useful for the plugin interface. I mostly care about having the formatter API public.

Comment (hukkinj1) [reply mdformat/renderer/__init__.py:121]: I'm not fully opposed to making any API public, I just think there should usually be a good use-case and hopefully also at least one user willing to use it.

Comment (chrisjsewell) [reply mdformat/renderer/__init__.py]: sure, its not a hill I want to die on lol; changed in 895d3106338c543d4267f5976af7978a3fdecaa6

Comment (hukkinj1) [reply mdformat/renderer/__init__.py]: :smile: nice thanks!

**Comment by hukkinj1** [comment]
Very nice, thanks @chrisjsewell !! Merged
